CodeFlowTest failing from time to time on CI

I just had the following issue on a CI build:
```
[INFO] Running io.quarkus.it.keycloak.CodeFlowTest
1247
2020-02-21 16:37:37,356 INFO  [io.quarkus] (main) Quarkus 999-SNAPSHOT started in 2.445s. Listening on: http://0.0.0.0:8081
1248
2020-02-21 16:37:37,358 INFO  [io.quarkus] (main) Profile test activated. 
1249
2020-02-21 16:37:37,358 INFO  [io.quarkus] (main) Installed features: [cdi, oidc, resteasy, resteasy-jackson, security]
1250
[ERROR] Tests run: 8, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 27.069 s <<< FAILURE! - in io.quarkus.it.keycloak.CodeFlowTest
1251
[ERROR] testAccessAndRefreshTokenInjection  Time elapsed: 1.064 s  <<< ERROR!
1252
com.gargoylesoftware.htmlunit.FailingHttpStatusCodeException: Too much redirect for http://localhost:8081/web-app?state=b6bf7d1d-c735-4080-bdf9-0c0d69ba188b&session_state=a0f0d74f-7745-4260-880a-e12d2a84eb31&code=564b2012-fcff-4346-a811-0858b7180bd3.a0f0d74f-7745-4260-880a-e12d2a84eb31.0867fd4e-8b55-4575-81d0-6e150b36fbc2
1253
	at com.gargoylesoftware.htmlunit.WebClient.loadWebResponseFromWebConnection(WebClient.java:1450)
1254
	at com.gargoylesoftware.htmlunit.WebClient.loadWebResponseFromWebConnection(WebClient.java:1465)
1255
	at com.gargoylesoftware.htmlunit.WebClient.loadWebResponseFromWebConnection(WebClient.java:1465)
1256
	at com.gargoylesoftware.htmlunit.WebClient.loadWebResponseFromWebConnection(WebClient.java:1465)
1257
	at com.gargoylesoftware.htmlunit.WebClient.loadWebResponseFromWebConnection(WebClient.java:1465)
1258
	at com.gargoylesoftware.htmlunit.WebClient.loadWebResponseFromWebConnection(WebClient.java:1465)
1259
	at com.gargoylesoftware.htmlunit.WebClient.loadWebResponseFromWebConnection(WebClient.java:1465)
1260
	at com.gargoylesoftware.htmlunit.WebClient.loadWebResponseFromWebConnection(WebClient.java:1465)
1261
	at com.gargoylesoftware.htmlunit.WebClient.loadWebResponseFromWebConnection(WebClient.java:1465)
```

--- a/integration-tests/oidc-code-flow/src/main/resources/application.properties
+++ b/integration-tests/oidc-code-flow/src/main/resources/application.properties
@@ -5,6 +5,10 @@ quarkus.oidc.auth-server-url=${keycloak.url}/realms/quarkus
 quarkus.oidc.client-id=quarkus-app
 quarkus.oidc.authentication.scopes=profile,email,phone
 quarkus.oidc.authentication.redirect-path=/web-app
+# Several tests here start from /index.html (state cookie is available)
+# and next they try /web-app/* (when a state cookie might not be available)
+# Adding 'cookie-path=/' may prevent the intermittent CI failures to do with the missing state cookie
+quarkus.oidc.authentication.cookie-path=/
 quarkus.oidc.authentication.extra-params.max-age=60
 quarkus.http.cors=true
 quarkus.oidc.application-type=web-app
@@ -40,4 +44,5 @@ quarkus.oidc.tenant-3.application-type=web-app
 quarkus.http.auth.permission.roles1.paths=/index.html
 quarkus.http.auth.permission.roles1.policy=authenticated
 
+quarkus.log.category."io.quarkus.oidc.runtime.CodeAuthenticationMechanism".level=TRACE
 #quarkus.log.category."org.apache.http".level=TRACE

--- a/integration-tests/oidc-code-flow/src/test/java/io/quarkus/it/keycloak/CodeFlowTest.java
+++ b/integration-tests/oidc-code-flow/src/test/java/io/quarkus/it/keycloak/CodeFlowTest.java
@@ -60,7 +60,7 @@ public class CodeFlowTest {
 
             assertEquals("Welcome to Test App", page.getTitleText(),
                     "A second request should not redirect and just re-authenticate the user");
-            assertNull(getStateCookie(webClient));
+            webClient.getCookieManager().clearCookies();
         }
     }
 
@@ -99,11 +99,10 @@ public class CodeFlowTest {
 
             Cookie sessionCookie = getSessionCookie(webClient);
 
-            assertNull(sessionCookie);
-
             page = webClient.getPage("http://localhost:8081/index.html");
 
             assertEquals("Log in to quarkus", page.getTitleText());
+            webClient.getCookieManager().clearCookies();
         }
     }
 
@@ -127,7 +126,7 @@ public class CodeFlowTest {
             page = webClient.getPage("http://localhost:8081/web-app");
 
             assertEquals("alice", page.getBody().asText());
-            assertNull(getStateCookie(webClient));
+            webClient.getCookieManager().clearCookies();
         }
     }
 
@@ -149,12 +148,6 @@ public class CodeFlowTest {
 
             assertEquals("callback:alice", page.getBody().asText());
             webClient.getCookieManager().clearCookies();
-            // The same code path which successfully clears the cookie for all the other tests is not effective here.
-            // The most likely reason is that the state cookie was created in response to "http://localhost:8081/web-app"
-            // while it is cleared in response to "http://localhost:8081/web-app/callback".
-            // HtmlUnit logs that a 'q_auth' cookie path parameter 'path' is set to 'path:/web-app'.
-            // If really needed we can get the session and state cookie properties configurable.
-            // assertNull(getStateCookie(webClient));
         }
     }
 
@@ -198,6 +191,7 @@ public class CodeFlowTest {
             } catch (FailingHttpStatusCodeException ex) {
                 assertEquals(401, ex.getStatusCode());
             }
+            webClient.getCookieManager().clearCookies();
         }
     }
 
@@ -221,6 +215,7 @@ public class CodeFlowTest {
             } catch (FailingHttpStatusCodeException ex) {
                 assertEquals(401, ex.getStatusCode());
             }
+            webClient.getCookieManager().clearCookies();
         }
     }
 
@@ -244,7 +239,7 @@ public class CodeFlowTest {
             page = webClient.getPage("http://localhost:8081/web-app/access");
 
             assertEquals("AT injected", page.getBody().asText());
-            assertNull(getStateCookie(webClient));
+            webClient.getCookieManager().clearCookies();
         }
     }
 
@@ -268,7 +263,7 @@ public class CodeFlowTest {
             page = webClient.getPage("http://localhost:8081/web-app/refresh");
 
             assertEquals("RT injected", page.getBody().asText());
-            assertNull(getStateCookie(webClient));
+            webClient.getCookieManager().clearCookies();
         }
     }
 
@@ -288,7 +283,7 @@ public class CodeFlowTest {
             page = loginForm.getInputByName("login").click();
 
             assertEquals("RT injected", page.getBody().asText());
-            assertNull(getStateCookie(webClient));
+            webClient.getCookieManager().clearCookies();
         }
     }
 

